web3.js/src/transaction.ts

```diff
@@ -327,17 +327,24 @@ export class Transaction {
       return this._message;
     }
 
-    const {nonceInfo} = this;
-    if (nonceInfo && this.instructions[0] != nonceInfo.nonceInstruction) {
-      this.recentBlockhash = nonceInfo.nonce;
-      this.instructions.unshift(nonceInfo.nonceInstruction);
+    let recentBlockhash;
+    let instructions: TransactionInstruction[];
+    if (this.nonceInfo) {
+      recentBlockhash = this.nonceInfo.nonce;
```

Comment:
_Always_ use the nonce as the recentBlockhash, but only in the scope of the `compileMessage` function.